grovedb/src/operations/proof.rs

```diff
@@ -135,41 +194,357 @@ impl GroveDb {
                     )),
                 }?;
 
-                if child_hash != last_root_hash {
+                if child_hash != expected_root_hash {
                     return Err(Error::InvalidProof("Bad path"));
                 }
 
-                last_root_hash = proof_result.0;
+                expected_root_hash = proof_result.0;
             } else {
                 break;
             }
             split_path = path_slice.split_last();
         }
 
-        let root_proof = proof_reader.read_proof(ROOT_PROOF)?;
+        // execute the root proof
+        let root_proof_bytes = proof_reader.read_proof_of_type(ProofType::RootProof.into())?;
+
+        // makes the assumption that 1 byte is enough to represent the root leaf count
+        // hence max of 255 root leaf keys
+        let root_leaf_count = proof_reader.read_byte()?;
 
-        let root_meta_data = proof_reader.read_to_end();
-        let root_index_usize = root_meta_data
+        let index_to_prove_as_bytes = proof_reader.read_to_end();
+        let index_to_prove_as_usize = index_to_prove_as_bytes
             .into_iter()
             .map(|index| index as usize)
             .collect::<Vec<usize>>();
 
-        let root_proof_terrible_name = match MerkleProof::<Sha256>::try_from(root_proof) {
+        let root_proof = match MerkleProof::<Sha256>::try_from(root_proof_bytes) {
             Ok(proof) => Ok(proof),
             Err(_) => Err(Error::InvalidProof("invalid proof element")),
         }?;
 
-        // TODO: Don't hard code the leave count
-        let root_hash = match root_proof_terrible_name.root(&root_index_usize, &[last_root_hash], 2)
-        {
+        let root_hash = match root_proof.root(
+            &index_to_prove_as_usize,
+            &[expected_root_hash],
+            root_leaf_count[0] as usize,
+        ) {
             Ok(hash) => Ok(hash),
             Err(_) => Err(Error::InvalidProof("Invalid proof element")),
         }?;
 
-        Ok((root_hash, result_set.result_set))
+        Ok((root_hash, result_set))
+    }
+
+    fn prove_subqueries(
+        db: &GroveDb,
+        proofs: &mut Vec<u8>,
+        path: Vec<&[u8]>,
+        query: PathQuery,
+        current_limit: &mut Option<u16>,
+        current_offset: &mut Option<u16>,
+    ) -> Result<(), Error> {
+        // there is a chance that the subquery key would lead to something that is not a
+        // tree same thing for the subquery itself
+        merk_optional_tx!(db.db, path.clone(), None, subtree, {
```

Comment:
`merk_optional_tx` macro is made to generate code that will use transactional context in case of `Some(transaction)` and a regular one in `None` case;

As transaction will be always `None`, the branch generated by macro for tx case is useless, so macro is not needed in such case

Howerver, it's a nice shortcut to create a storage context and open a Merk on it, but it still doesn't feel right

Reply:
added a todo to fix this in the latest branch